### PR TITLE
Rework cross-signing login flow

### PR DESCRIPTION
**Pull request by ara4n**
[This thread](https://twitter.com/popey/status/1366308809287860230) really struck home (I assume it's one of the reasons we lost Canonical), and the fact that we've failed to budget time to iterate and improve the deeply suboptimal cross-signing login flows since launching it at FOSDEM 2020 is very frustrating.  So I went on a hack this evening to apply a series of targeted strikes to fix it:

 * When logging in to an existing account and setting up encryption, you now get an explicit "verify using another session" button rather expecting the user to figure out they need to initiate verification from another session.  Likewise when you hit the "Verify this session" toast.
 * We don't show the dialog at all if the user doesn't have an MSK to self-verify, and has no devices to verify against.
 * The "new login" notification now doesn't prompt you to verify, which rarely worked given the new login would likely be stuck in the middle of an initial sync and not able to be verified.  Instead it prompts you to check your devices (taking you to the place you'd kick out the new login if you didn't recognise it)
 * Adds an explicit link to edit your devices from the verification UI on your own UserInfo
 * If you accidentally click in the background it now cancels the process correctly
 * Rename the annoying "Review your logins!" prompt to actually explain *why* it's asking you to review your logins.
 * We now report IPs of new logins and self-verification attempts

This PR obviously should be split into 7 or so small PRs, but for expediency i've bundled them together here.  *However*, the commits are nicely structured so they can be reviewed one by one.

Depends on https://github.com/matrix-org/matrix-js-sdk/pull/1632

---

@jryans: I've collected some open issues that I believe are solved by this.

Fixes https://github.com/vector-im/element-web/issues/16252
Fixes https://github.com/vector-im/element-web/issues/15877
Fixes https://github.com/vector-im/element-web/issues/15672
Fixes https://github.com/vector-im/element-web/issues/15429
Fixes https://github.com/vector-im/element-web/issues/13289
Fixes https://github.com/vector-im/element-web/issues/13510

### CODE REVIEW

File: src/components/views/toasts/VerificationRequestToast.tsx
Comment (jryans) [inline src/components/views/toasts/VerificationRequestToast.tsx:128]: Hmm, should the request _always_ be cancelled on finish? If yes, I am puzzled why you've changed to cancel manually at each call site instead of within the dialog...

File: src/components/views/toasts/VerificationRequestToast.tsx
Comment (jryans) [inline src/components/views/toasts/VerificationRequestToast.tsx:41]: I'd suggest spelling this `ip`, since we use e.g. `uri` for URIs and other similar acronyms.

File: src/components/views/toasts/VerificationRequestToast.tsx
Comment (jryans) [inline src/components/views/toasts/VerificationRequestToast.tsx:144]: Hmm, having two "from"s seems a bit unfortunate, perhaps the IP could go inside the parens with the device ID...?

File: src/stores/SetupEncryptionStore.js
Comment (jryans) [inline src/stores/SetupEncryptionStore.js:52]: Nit: Remove this extra blank line

File: src/stores/SetupEncryptionStore.js
Comment (jryans) [inline src/stores/SetupEncryptionStore.js:100]: As in the verification prompt appears briefly before we realise there are no devices...? Perhaps we can use a "loading" phase first, then we go to either intro or finished based on the results?

File: src/toasts/UnverifiedSessionToast.ts
Comment (jryans) [inline src/toasts/UnverifiedSessionToast.ts]: Perhaps spell it `ip` here.

File: src/toasts/UnverifiedSessionToast.ts
Comment (jryans) [inline src/toasts/UnverifiedSessionToast.ts:53]: If we agree the other label should move the IP in the parens with the device ID, perhaps do the same here as well.

File: src/toasts/UnverifiedSessionToast.ts
Comment (jryans) [inline src/toasts/UnverifiedSessionToast.ts:37]: Hmm, have you added a way to start verification from settings...?

File: src/components/structures/auth/SetupEncryptionBody.js
Comment (jryans) [inline src/components/structures/auth/SetupEncryptionBody.js:155]: If all 3 buttons are shown, they seem to wrap to 2 lines, at least in Firefox:

image

**Review by jryans — CHANGES_REQUESTED** [review]
Overall, it seems like a more reliable workflow, thanks! 😄 

I've added a few details we should work out before merging.

File: src/components/views/toasts/VerificationRequestToast.tsx
Comment (t3chguy) [reply src/components/views/toasts/VerificationRequestToast.tsx:144]: "from $device at $ip" maybe

File: src/components/views/toasts/VerificationRequestToast.tsx
Comment (ara4n) [reply src/components/views/toasts/VerificationRequestToast.tsx:128]: i think the answer is "yes" in that if the modal finishes prematurely then you want to cancel the request in flight, and if it doesn't finish prematurely then the request has already been and gone and it doesn't matter if you try to cancel it.  Originally the VerificationRequestDialog had its own onFinished which always cancelled the request, but there was no way for Modal.createTrackedDialog() to call the inner onFinished if the modal is cancelled at the Modal layer (i.e. by clicking in the background, which is all too easy to do when you click on a toast; nothing seems to happen; you click again on the next toast down; you cancel the verif modal).  So, rather than trying to figure out a way to invoke the inner onFinished() from the Modal, instead I removed it given the request.cancel() is essentially one line of code and put it in the outer onFinished().

File: src/toasts/UnverifiedSessionToast.ts
Comment (ara4n) [reply src/toasts/UnverifiedSessionToast.ts:37]: nope, there was one already.

Comment (jryans) [reply src/toasts/UnverifiedSessionToast.ts:37]: Wait, where...? It's not in the session list AFAICT... Anyway, the new flow is still okay without it, I just wanted to be sure of your intention here.

Comment (ara4n) [reply src/toasts/UnverifiedSessionToast.ts:37]: there's a link to your memberinfo from the top of the device section.

File: src/components/structures/auth/SetupEncryptionBody.js
Comment (ara4n) [reply src/components/structures/auth/SetupEncryptionBody.js]: fixed (and improved the wrap for other translations): 
Screenshot 2021-03-08 at 17 59 17

File: src/stores/SetupEncryptionStore.js
Comment (ara4n) [reply src/stores/SetupEncryptionStore.js:100]: done.

**Comment by ara4n** [comment]
@jryans PTAL

File: src/components/views/toasts/VerificationRequestToast.tsx
Comment (jryans) [reply src/components/views/toasts/VerificationRequestToast.tsx:128]: Ah I see, makes sense then.

File: src/toasts/UnverifiedSessionToast.ts
Comment (jryans) [reply src/toasts/UnverifiedSessionToast.ts:37]: Ah right, it's a bit indirect, but anyway it's at least possible to jump where needed.

**Comment by jryans** [comment]
Adding Design review as well, so they can take a look at these workflow changes before merging.

**Review by jryans — APPROVED** [review]
Looks good to me code wise, with one tiny cleanup needed. I have added this PR to the Design queue as well.

**Comment by jryans** [comment]
It likely to take another day anyway given Design review, but I would suggest we wait until after the RC is cut tomorrow to merge this, so we can have maximal bake time on develop.

**Review by nadonomy — APPROVED** [review]
Approving to remove the unnecessary paths. It would be good to merge this in to test more generally and test stability on develop, but @jryans we should park some time Mar 22-24 ahead of the next RC to tweak some of the copy. I'll book something in our calendars as a reminder.

**Pull request merged**
